FogCreek/Glitch-Community
# fern-street: implement pagination and filtering for collections

## Links
* [fern-street.glitch.me](https://fern-street.glitch.me)
* [Ticket](https://github.com/FogCreek/Glitch-Community-Work/issues/182)

## GIF/Screenshots:
![collections-pagination](https://user-images.githubusercontent.com/7584833/61547360-39139600-aa19-11e9-9e34-65a581dbd97c.gif)

<div>
<img width="450" src="https://user-images.githubusercontent.com/7584833/61551382-07073180-aa23-11e9-8c20-ba7d40032be3.png" />
<img width="350" src="https://user-images.githubusercontent.com/7584833/61551331-dd4e0a80-aa22-11e9-9000-36e4acc299a0.png"/>
</div>

## Changes:
Implement pagination for collections on teams and user pages.

We previously displayed all collections and only fetched/rendered the projects as they came into view. This was becoming unsustainable on users/teams with hundreds of collections, [@tagging](https://glitch.com/@tagging), for example (⚠that link may crash your browser)

* Factor out FilterController and PaginationController into reusable components, use them for collections and projects list on team/user pages
  * Add `matchFn` prop to FilterController for customizing match criteria
  * Add `fetchDataOptimistically` prop to PaginationController, for optimistically making API requests for the upcoming page
* Add `showLoader` prop to `<CollectionItem>` to hide loaders when `fetchDataOptimistically` is used

## How To Test:
- Confirm that pagination and filtering still works for *projects* on team and user pages
- Check that collection pagination works for team and user pages [@glitch](https://fern-street.glitch.me) is a good page to test
    - Can you filter collections?
    - Can you move forward and backward between pages?
    - Does clicking the "Show all" button work? And does it show the loader for collections that haven't been loaded yet?
